linkerd/lock/src/shared.rs

```diff
@@ -0,0 +1,166 @@
+use self::waiter::Notify;
+pub(crate) use self::waiter::Wait;
+use crate::error::Error;
+use futures::{Async, Poll};
+use std::sync::Arc;
+use tracing::trace;
+
+/// The shared state between one or more Lock instances.
+///
+/// When multiple lock instances are contending for the inner value, waiters are
+/// stored in a LIFO stack. This is done to bias for latency instead of fairness
+pub(crate) struct Shared<T> {
+    state: State<T>,
+    waiters: Vec<Notify>,
+}
+
+enum State<T> {
+    /// A Lock is holding the value.
+    Claimed,
+
+    /// The inner value is available.
+    Unclaimed(T),
+
+    /// The lock has failed.
+    Failed(Arc<Error>),
+}
+
+// === impl Shared ===
+
+impl<T> Shared<T> {
+    pub fn new(value: T) -> Self {
+        Self {
+            waiters: Vec::new(),
+            state: State::Unclaimed(value),
+        }
+    }
+
+    /// Try to claim a value without registering a waiter.
+    pub fn try_acquire(&mut self) -> Result<Option<T>, Arc<Error>> {
+        match std::mem::replace(&mut self.state, State::Claimed) {
+            // This lock has acquired the value.
+            State::Unclaimed(v) => {
+                trace!("acquired");
+                Ok(Some(v))
+            }
+            // The value is already claimed by a lock.
+            State::Claimed => Ok(None),
+            // The locks failed, so reset the state immediately so that all
+            // instances may be notified.
+            State::Failed(error) => {
+                self.state = State::Failed(error.clone());
+                Err(error)
+            }
+        }
+    }
+
+    /// Try to acquire a value or register the given waiter to be notified when
+    /// the lock is available.
+    pub fn poll_acquire(&mut self, wait: &Wait) -> Poll<T, Arc<Error>> {
+        match self.try_acquire() {
+            Ok(Some(svc)) => Ok(Async::Ready(svc)),
+            Ok(None) => {
+                // Register the current task to be notified.
+                wait.register();
+                // Register the waiter's notify handle if one isn't already registered.
+                if let Some(notify) = wait.get_notify() {
+                    trace!("Registering waiter");
+                    self.waiters.push(notify);
+                }
+                debug_assert!(wait.is_waiting());
+                Ok(Async::NotReady)
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    pub fn release_and_notify(&mut self, value: T) {
+        trace!(waiters = self.waiters.len(), "Releasing");
+        debug_assert!(match self.state {
```

Comment:
This assertion is to catch misuse of `Shared`? Or sanity-checking internals?

If it's to catch misuse, I prefer a regular `assert!`, since there may some cases it *is* misused and only noticed under load. If it's for internals, ignore me :D